### ISSUE TITLE
After upgrading to 0.25, unknown protocol error. 

### ISSUE THREAD

**Issue by hannesroos** [metabase/metabase#5585]
After updating Metabase to 0.25, it does not run, the following error is shown in the log

`07-26 10:54:28 WARN server.HttpChannel :: /
java.net.MalformedURLException: unknown protocol: my.domain.name
        at java.net.URL.<init>(URL.java:600)
        at java.net.URL.<init>(URL.java:490)
        at java.net.URL.<init>(URL.java:439)
        at clojure.java.io$fn__9408.invokeStatic(io.clj:44)
        at clojure.java.io$fn__9408.invoke(io.clj:44)
        at clojure.java.io$fn__9379$G__9374__9384.invoke(io.clj:35)
        at metabase.routes$base_href.invokeStatic(routes.clj:20)
        at metabase.routes$base_href.invoke(routes.clj:19)
        at metabase.routes$entrypoint.invokeStatic(routes.clj:39)
        at metabase.routes$entrypoint.invoke(routes.clj:34)
        at clojure.core$partial$fn__4761.invoke(core.clj:2522)
        at compojure.response$fn__206.invokeStatic(response.clj:33)
        at compojure.response$fn__206.invoke(response.clj:21)
        at compojure.response$fn__179$G__174__186.invoke(response.clj:6)
        at compojure.core$make_route$fn__1561.invoke(core.clj:135)
        at compojure.core$wrap_route_middleware$fn__1554.invoke(core.clj:122)
        at compojure.core$wrap_route_info$fn__1558.invoke(core.clj:126)
        at compojure.core$if_route$fn__1516.invoke(core.clj:45)
        at compojure.core$if_method$fn__1506.invoke(core.clj:27)
        at compojure.core$routing$fn__1568.invoke(core.clj:151)
        at clojure.core$some.invokeStatic(core.clj:2592)
        at clojure.core$some.invoke(core.clj:2583)
        at compojure.core$routing.invokeStatic(core.clj:151)
        at compojure.core$routing.doInvoke(core.clj:148)
        at clojure.lang.RestFn.applyTo(RestFn.java:139)
        at clojure.core$apply.invokeStatic(core.clj:648)
        at clojure.core$apply.invoke(core.clj:641)
        at compojure.core$routes$fn__1572.invoke(core.clj:156)
        at clojure.lang.Var.invoke(Var.java:379)
        at metabase.middleware$log_api_call$fn__29059.invoke(middleware.clj:328)
        at metabase.middleware$add_security_headers$fn__29009.invoke(middleware.clj:245)
        at ring.middleware.json$wrap_json_body$fn__41257.invoke(json.clj:46)
        at ring.middleware.json$wrap_json_response$fn__41274.invoke(json.clj:87)
        at ring.middleware.keyword_params$wrap_keyword_params$fn__41302.invoke(keyword_params.clj:36)
        at ring.middleware.params$wrap_params$fn__41350.invoke(params.clj:67)
        at metabase.middleware$bind_current_user$fn__28964.invoke(middleware.clj:142)
        at clojure.core$comp$fn__4727.invoke(core.clj:2460)
        at clojure.core$comp$fn__4727.invoke(core.clj:2460)
        at clojure.core$comp$fn__4727.invoke(core.clj:2460)
        at metabase.middleware$maybe_set_site_url$fn__29013.invoke(middleware.clj:268)
        at ring.middleware.cookies$wrap_cookies$fn__41200.invoke(cookies.clj:175)
        at ring.middleware.session$wrap_session$fn__41441.invoke(session.clj:108)
        at ring.middleware.gzip$wrap_gzip$fn__41231.invoke(gzip.clj:65)
        at ring.adapter.jetty$proxy_handler$fn__41077.invoke(jetty.clj:25)
        at ring.adapter.jetty.proxy$org.eclipse.jetty.server.handler.AbstractHandler$ff19274a.handle(Unknown Source)
        at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:97)
        at org.eclipse.jetty.server.Server.handle(Server.java:499)
        at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:311)
        at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:257)
        at org.eclipse.jetty.io.AbstractConnection$2.run(AbstractConnection.java:544)
        at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:635)
        at org.eclipse.jetty.util.thread.QueuedThreadPool$3.run(QueuedThreadPool.java:555)
        at java.lang.Thread.run(Thread.java:745)`

With my.domain.name referring to the URL of the server I have installed it on. 

When I try to open it in browser (my.domain.name:3000), the following error appears

`HTTP ERROR: 500
Problem accessing /. Reason:
unknown protocol: my.domain.name
Powered by Jetty://`

When I downgrade to old version (0.24.2), it continues to work. And when I create a new instance in a new folder, it works as well.

I am running Ubuntu Server 14.04, and use the jar file of Metabase. `java -version` shows

`openjdk version "1.8.0_111"
OpenJDK Runtime Environment (build 1.8.0_111-8u111-b14-3~14.04.1-b14)
OpenJDK 64-Bit Server VM (build 25.111-b14, mixed mode)`

Thanks in advance!

**Comment by cwbeck** [comment]
Same issue...

```
:/# docker logs -f metabase-migrate
07-26 11:33:02 INFO metabase.util :: Loading Metabase...
07-26 11:33:10 INFO util.encryption :: DB details encryption is DISABLED for this Metabase instance. 🔓
07-26 11:33:21 INFO metabase.core :: Starting Metabase in STANDALONE mode
07-26 11:33:21 INFO metabase.core :: Launching Embedded Jetty Webserver with config:
 {:port 3000, :host "91a1f0c82114"}

07-26 11:33:21 INFO metabase.core :: Starting Metabase version v0.25.0 (b1316d5 release-0.25.0) ...
07-26 11:33:21 INFO metabase.core :: System timezone is 'GMT' ...
07-26 11:33:21 DEBUG metabase.driver :: Registered driver :bigquery 🚚
07-26 11:33:21 DEBUG metabase.driver :: Registered driver :crate 🚚
07-26 11:33:21 DEBUG metabase.driver :: Registered driver :druid 🚚
07-26 11:33:21 WARN metabase.driver :: No -init-driver function found for 'metabase.driver.google'
07-26 11:33:22 DEBUG metabase.driver :: Registered driver :googleanalytics 🚚
07-26 11:33:22 DEBUG metabase.driver :: Registered driver :h2 🚚
07-26 11:33:22 DEBUG metabase.driver :: Registered driver :mongo 🚚
07-26 11:33:22 DEBUG metabase.driver :: Registered driver :mysql 🚚
07-26 11:33:22 DEBUG metabase.driver :: Registered driver :postgres 🚚
07-26 11:33:22 DEBUG metabase.driver :: Registered driver :presto 🚚
07-26 11:33:22 DEBUG metabase.driver :: Registered driver :redshift 🚚
07-26 11:33:22 DEBUG metabase.driver :: Registered driver :sqlite 🚚
07-26 11:33:22 DEBUG metabase.driver :: Registered driver :sqlserver 🚚
07-26 11:33:22 INFO metabase.core :: Setting up and migrating Metabase DB. Please sit tight, this may take a minute...
07-26 11:33:22 INFO metabase.db :: Verifying h2 Database Connection ...
07-26 11:33:23 INFO metabase.db :: Verify Database Connection ...  ✅
07-26 11:33:23 INFO metabase.db :: Running Database Migrations...
07-26 11:33:23 INFO metabase.db :: Setting up Liquibase...
07-26 11:33:23 INFO metabase.db :: Liquibase is ready.
07-26 11:33:23 INFO metabase.db :: Checking if Database has unrun migrations...
07-26 11:33:27 INFO metabase.db :: Database has unrun migrations. Waiting for migration lock to be cleared...
07-26 11:33:27 INFO metabase.db :: Migration lock is cleared. Running migrations...
07-26 11:33:31 INFO metabase.db :: Database Migrations Current ...  ✅
com.mchange.v2.cfg.DelayedLogItem [ level -> FINE, text -> "The configuration file for resource identifier 'hocon:/reference,/application,/c3p0,/' could not be found. Skipping.", exception -> null]
07-26 11:33:31 INFO db.migrations :: Running all necessary data migrations, this may take a minute.
07-26 11:33:32 INFO db.migrations :: Finished running data migrations.
07-26 11:33:32 INFO metabase.events :: Starting events listener: metabase.events.activity-feed 👂
07-26 11:33:32 INFO metabase.events :: Starting events listener: metabase.events.dependencies 👂
07-26 11:33:32 INFO metabase.events :: Starting events listener: metabase.events.driver-notifications 👂
07-26 11:33:32 INFO metabase.events :: Starting events listener: metabase.events.last-login 👂
07-26 11:33:32 INFO metabase.events :: Starting events listener: metabase.events.metabot-lifecycle 👂
07-26 11:33:32 INFO metabase.events :: Starting events listener: metabase.events.notifications 👂
07-26 11:33:32 INFO metabase.events :: Starting events listener: metabase.events.revision 👂
07-26 11:33:32 INFO metabase.events :: Starting events listener: metabase.events.sync-database 👂
07-26 11:33:32 INFO metabase.events :: Starting events listener: metabase.events.view-log 👂
07-26 11:33:32 INFO metabase.task :: Loading tasks namespace: metabase.task.follow-up-emails 📆
07-26 11:33:32 INFO metabase.task :: Loading tasks namespace: metabase.task.send-anonymous-stats 📆
07-26 11:33:32 INFO metabase.task :: Loading tasks namespace: metabase.task.send-pulses 📆
07-26 11:33:32 INFO metabase.task :: Loading tasks namespace: metabase.task.sync-databases 📆
07-26 11:33:32 INFO metabase.task :: Loading tasks namespace: metabase.task.upgrade-checks 📆
07-26 11:33:32 INFO metabase.core :: Metabase Initialization COMPLETE
07-26 11:33:46 WARN server.HttpChannel :: /
java.net.MalformedURLException: no protocol: mb.venatusmedia.com
	at java.net.URL.<init>(URL.java:593)
	at java.net.URL.<init>(URL.java:490)
	at java.net.URL.<init>(URL.java:439)
	at clojure.java.io$fn__9408.invokeStatic(io.clj:44)
	at clojure.java.io$fn__9408.invoke(io.clj:44)
	at clojure.java.io$fn__9379$G__9374__9384.invoke(io.clj:35)
	at metabase.routes$base_href.invokeStatic(routes.clj:20)
	at metabase.routes$base_href.invoke(routes.clj:19)
	at metabase.routes$entrypoint.invokeStatic(routes.clj:39)
	at metabase.routes$entrypoint.invoke(routes.clj:34)
	at clojure.core$partial$fn__4761.invoke(core.clj:2522)
	at compojure.response$fn__206.invokeStatic(response.clj:33)
	at compojure.response$fn__206.invoke(response.clj:21)
	at compojure.response$fn__179$G__174__186.invoke(response.clj:6)
	at compojure.core$make_route$fn__1561.invoke(core.clj:135)
	at compojure.core$wrap_route_middleware$fn__1554.invoke(core.clj:122)
	at compojure.core$wrap_route_info$fn__1558.invoke(core.clj:126)
	at compojure.core$if_route$fn__1516.invoke(core.clj:45)
	at compojure.core$if_method$fn__1506.invoke(core.clj:27)
	at compojure.core$routing$fn__1568.invoke(core.clj:151)
	at clojure.core$some.invokeStatic(core.clj:2592)
	at clojure.core$some.invoke(core.clj:2583)
	at compojure.core$routing.invokeStatic(core.clj:151)
	at compojure.core$routing.doInvoke(core.clj:148)
	at clojure.lang.RestFn.applyTo(RestFn.java:139)
	at clojure.core$apply.invokeStatic(core.clj:648)
	at clojure.core$apply.invoke(core.clj:641)
	at compojure.core$routes$fn__1572.invoke(core.clj:156)
	at clojure.lang.Var.invoke(Var.java:379)
	at metabase.middleware$log_api_call$fn__29059.invoke(middleware.clj:328)
	at metabase.middleware$add_security_headers$fn__29009.invoke(middleware.clj:245)
	at ring.middleware.json$wrap_json_body$fn__41257.invoke(json.clj:46)
	at ring.middleware.json$wrap_json_response$fn__41274.invoke(json.clj:87)
	at ring.middleware.keyword_params$wrap_keyword_params$fn__41302.invoke(keyword_params.clj:36)
	at ring.middleware.params$wrap_params$fn__41350.invoke(params.clj:67)
	at metabase.middleware$bind_current_user$fn__28964.invoke(middleware.clj:143)
	at clojure.core$comp$fn__4727.invoke(core.clj:2460)
	at clojure.core$comp$fn__4727.invoke(core.clj:2460)
	at clojure.core$comp$fn__4727.invoke(core.clj:2460)
	at metabase.middleware$maybe_set_site_url$fn__29013.invoke(middleware.clj:268)
	at ring.middleware.cookies$wrap_cookies$fn__41200.invoke(cookies.clj:175)
	at ring.middleware.session$wrap_session$fn__41441.invoke(session.clj:108)
	at ring.middleware.gzip$wrap_gzip$fn__41231.invoke(gzip.clj:65)
	at ring.adapter.jetty$proxy_handler$fn__41077.invoke(jetty.clj:25)
	at ring.adapter.jetty.proxy$org.eclipse.jetty.server.handler.AbstractHandler$ff19274a.handle(Unknown Source)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:97)
	at org.eclipse.jetty.server.Server.handle(Server.java:499)
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:311)
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:257)
	at org.eclipse.jetty.io.AbstractConnection$2.run(AbstractConnection.java:544)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:635)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$3.run(QueuedThreadPool.java:555)
	at java.lang.Thread.run(Thread.java:745)
```

**Comment by senior** [comment]
I was able to reproduce this and have a workaround.

To reproduce it:
- Launch Metabase 0.23.0 with a fresh install (no existing database)
- Go through the normal/basic Metabase setup steps, no need to connect to a DB
- Open the admin panel, click General, specify a site url without a protocol (i.e. "metabase.localdomain:3000")
- Shutdown 0.23.0
- Launch 0.24.2 - observe it's working - shutdown
- Launch 0.25.0 - hit the root url (i.e. "http://metabase.localdomain:3000/")
- Fails with MalformedURLException

To workaround this:
- Downgrade metabase to something before 0.25.0 (I used 0.24.2)
- Open the admin panel, click "General", and add a leading http or https to your URL (whichever is appropriate) for me this was "http://metabase.localdomain:3000".
- Shutdown the older version of metabase
- Launch 0.25.0

**Comment by salsakran** [comment]
We should seriously consider breaking up the site url into a site url and a protocol. IMO a string setting that contains two non-optional substrings shouldn't be stored as a single string. Schema (whether the setting type or SQL schema) exist for this reason.

Additionally, given that it's 2017, we should assume that most people that don't include a protocol are using http.

**Comment by salsakran** [comment]
Also, did you find the specific thing that expected a protocol and didn't get it and why it crept in?

**Comment by senior** [comment]
@salsakran I did

On every request, we populate the base_href (here)[https://github.com/metabase/metabase/commit/a8232b203ae987b5644f1cb9e6c30279519fb2f1#diff-c0ba3a75ffd69cedc7a144a6cbc44242R23]. That is new with this release. That calls
```
(io/as-url (public-settings/site-url))
```
*as-url* won't tolerate an invalid URL (such as one missing HTTP).

**Comment by senior** [comment]
Just to clarify a bit, the following has to be true for the bug to affect your install:

1. You have setup a site-url in the General section of the Admin panel that does NOT include http:// or https:// in the URL
2. You had this set when on version 0.23.0 (either specified using 0.23.0, or upgraded to 0.23.0)
3. You then upgrade to 0.25.0 (any migrations in-between 0.23.0 and 0.25.0 don't affect this issue)

The reason that this bug affects such a specific upgrade path is that 0.23.0 included migration code [here](https://github.com/metabase/metabase/blob/master/src/metabase/db/migrations.clj#L328) that updated the site-url. If that migration code runs on version 0.23.0 and you already had a site-url that doesn't have the protocol, it gets stored without it. Any version after that (such as 0.23.1) has code that will [automatically add the leading http/https](https://github.com/metabase/metabase/commit/e44f4b9ed54b2c17c960704dd77a2ddf92d0b705#diff-cae2134f5109fbdcdc7a17fa5b63d74fR32). Running that same migration code on a later version will enforce that the http/https is present. Once the migration has ran on 0.23.0, it won't run again on a future version to correct itself.

**Comment by salsakran** [comment]
this was fixed in #5587 which we'll be releasing in 0.25.1 very shortly

**Comment by salsakran** [comment]
@cwbeck  + @hannesroos 

We just shipped 0.25.1 which fixes this problem. Sorry for the hassle!
